gcp/src/main/java/org/apache/iceberg/gcp/gcs/OAuth2RefreshCredentialsHandler.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.gcp.gcs;
+
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.OAuth2CredentialsWithRefresh;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.iceberg.gcp.GCPProperties;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.rest.ErrorHandlers;
+import org.apache.iceberg.rest.HTTPClient;
+import org.apache.iceberg.rest.RESTClient;
+import org.apache.iceberg.rest.auth.OAuth2Properties;
+import org.apache.iceberg.rest.auth.OAuth2Util;
+import org.apache.iceberg.rest.credentials.Credential;
+import org.apache.iceberg.rest.responses.LoadCredentialsResponse;
+
+public class OAuth2RefreshCredentialsHandler
+    implements OAuth2CredentialsWithRefresh.OAuth2RefreshHandler {
+  private volatile HTTPClient client;
+  private final Map<String, String> properties;
+
+  private OAuth2RefreshCredentialsHandler(Map<String, String> properties) {
+    Preconditions.checkArgument(
+        null != properties.get(GCPProperties.GCS_OAUTH2_REFRESH_CREDENTIALS_ENDPOINT),
+        "Invalid credentials endpoint: null");
+    this.properties = properties;
+  }
+
+  @Override
+  public AccessToken refreshAccessToken() {
+    LoadCredentialsResponse response =
+        httpClient()
+            .get(
+                properties.get(GCPProperties.GCS_OAUTH2_REFRESH_CREDENTIALS_ENDPOINT),
+                null,
+                LoadCredentialsResponse.class,
+                OAuth2Util.authHeaders(properties.get(OAuth2Properties.TOKEN)),
+                ErrorHandlers.defaultErrorHandler());
+
+    List<Credential> gcsCredentials =
+        response.credentials().stream()
+            .filter(c -> c.prefix().startsWith("gs"))
+            .collect(Collectors.toList());
+
+    Preconditions.checkState(!gcsCredentials.isEmpty(), "Invalid GCS Credentials: empty");
+
+    Optional<Credential> credentialWithPrefix =
+        gcsCredentials.stream().max(Comparator.comparingInt(c -> c.prefix().length()));
```

Comment:
Just want to call out that we have the [same issue as with s3](https://github.com/apache/iceberg/pull/11389/files#r1815498305).